src/client/paired.rs

```diff
@@ -196,14 +196,85 @@ pub struct PairedConnection {
 
 async fn inner_conn_fn(
     addr: SocketAddr,
+    username: Option<String>,
+    password: Option<String>,
 ) -> Result<mpsc::UnboundedSender<SendPayload>, error::Error> {
     let connection = connect(&addr).await?;
     let (out_tx, out_rx) = mpsc::unbounded();
     let paired_connection_inner = PairedConnectionInner::new(connection, out_rx);
     tokio::spawn(paired_connection_inner);
+
+    if let Some(password) = password {
+        let mut args: Vec<resp::RespValue> = vec!["AUTH".into()];
+        if let Some(username) = username {
+            args.push(username.into())
+        }
+        args.push(password.into());
+        let (tx, rx) = oneshot::channel();
+        out_tx.unbounded_send((resp::RespValue::Array(args), tx))?;
+        match rx.await {
+            Ok(resp::RespValue::SimpleString(_)) => (),
+            Ok(resp::RespValue::Error(e)) => return Err(error::Error::Remote(e)),
+            Ok(_) => return Err(error::internal("Unexpected response to AUTH")),
+            Err(_) => {
+                return Err(error::Error::Internal(
+                    "Connection canceled while authenticating".to_string(),
+                ))
+            }
+        }
+    }
     Ok(out_tx)
 }
 
+/// Connection builder for construction a PairedConnection
+pub struct PairedConnectionBuilder {
+    addr: SocketAddr,
+    username: Option<String>,
+    password: Option<String>,
+}
+
+impl PairedConnectionBuilder {
+    pub fn new<A: ToSocketAddrs>(addr: A) -> Result<Self, error::Error> {
+        Ok(Self {
+            addr: addr.to_socket_addrs()?.next().ok_or_else(|| {
+                error::Error::Connection(error::ConnectionReason::ConnectionFailed)
+            })?,
+            username: None,
+            password: None,
+        })
+    }
+
+    /// Set the username used when connecting
+    pub fn password(&mut self, password: String) -> &mut Self {
+        self.password = Some(password);
+        self
+    }
+
+    /// Set the password used when connecting
+    pub fn username(&mut self, username: String) -> &mut Self {
+        self.username = Some(username);
+        self
+    }
+
+    pub async fn connect(&self) -> Result<PairedConnection, error::Error> {
+        let addr = self.addr;
+        let username = self.username.to_owned();
+        let password = self.password.to_owned();
+        let reconnecting_con = reconnect(
+            |con: &mpsc::UnboundedSender<SendPayload>, act| {
+                con.unbounded_send(act).map_err(|e| e.into())
+            },
+            move || {
+                let con_f = inner_conn_fn(addr, username.clone(), password.clone());
```

Comment:
The `.clone()` calls here could be avoided by allowing the `inner_conn_fn` to accept `Option<&str>` instead.

It won't make any difference in terms of efficiency, as they're cloned internally to take ownership anyway as the connection is asynchronous with a different lifecycle.

Reply:
This will not work as the lifetime of the returned `` outlives the username and password variables. 

The outer `to_owned` (could be clone), are needed unless the connect method should eat the builder. I my code I need two Redis connections (one for sending, and one for BLPOP), so for me it is most convenient to be able to call connect twice. These clones are moved into the `c` closure. 

The second `clone` is needed since the `Box::new(Box::pin(con_f))` may outlive the `c` closure, so we cannot use `&str`. Since the `c` closure in not `FnOnce` we cannot move the username and password into `con_f`. There might be some lifetime magic I could do to enforce that `c` must outlive `Box::new(Box::pin(con_f))`, but I have not been able to fight my way through that. And in the end I am not sure the complications would be worth it.